pkg/entry/mongo/document.go

```diff
@@ -16,42 +16,124 @@ import (
 
 const TimeFormat = time.RFC3339Nano
 
-type Document struct {
+type LogEntry interface {
+	Populate(ctx context.Context, record map[interface{}]interface{}) error
+	CollectionName() string
+	SaveTo(collection *mgo.Collection) error
+	GetID() bson.ObjectId
+}
+
+type LogDocument struct {
 	Id             bson.ObjectId `bson:"_id,omitempty"`
 	Log            string        `bson:"log"`
 	Stream         string        `bson:"stream"`
 	Time           string        `bson:"time"`
-	JobExecutionId string        `bson:"job_execution_id"`
 	ProjectId      string        `bson:"project_id"`
 	Customer       string        `bson:"customer"`
 	PlatformId     string        `bson:"platform_id"`
 }
 
-func Convert(ctx context.Context, ts time.Time, record map[interface{}]interface{}) (*Document, error) {
-	doc := &Document{}
+func (d *LogDocument) GetID() bson.ObjectId {
+	return d.Id
+}
 
-	if !ts.IsZero() {
-		doc.Time = ts.Format(TimeFormat)
-	}
+type JobLogDocument struct {
+	LogDocument 				 `bson:",inline"`
+	JobExecutionId string        `bson:"job_execution_id"`
+}
 
-	if err := doc.Populate(ctx, record); err != nil {
-		return nil, fmt.Errorf("populate document: %w", err)
-	}
+type AppLogDocument struct {
+	LogDocument 				 `bson:",inline"`
+	AppExecutionId string        `bson:"app_execution_id"`
+	AppId 		   string        `bson:"app_id"`
+	ContainerId    string        `bson:"container_id"`
+}
+
+func Convert(ctx context.Context, ts time.Time, record map[interface{}]interface{}) (LogEntry, error) {
+	if isJobLog(record) {
+		doc := &JobLogDocument{}
+
+		if !ts.IsZero() {
+			doc.Time = ts.Format(TimeFormat)
+		}
+
+		if err := doc.Populate(ctx, record); err != nil {
+			return nil, fmt.Errorf("populate document: %w", err)
+		}
+
+		return doc, nil
+	} else {
+		doc := &AppLogDocument{}
+
+		if !ts.IsZero() {
+			doc.Time = ts.Format(TimeFormat)
+		}
+
+		if err := doc.Populate(ctx, record); err != nil {
+			return nil, fmt.Errorf("populate document: %w", err)
+		}
 
```

Comment:
I suggest to move the common part out of the if:

```suggestion
func Convert(ctx context.Context, ts time.Time, record map[interface{}]interface{}) (LogEntry, error) {
	var doc LogEntry

	if isJobLog(record) {
		doc = &JobLogDocument{}
	} else {
		doc = &AppLogDocument{}
	}

	if !ts.IsZero() {
		doc.Time = ts.Format(TimeFormat)
	}

	if err := doc.Populate(ctx, record); err != nil {
		return nil, fmt.Errorf("populate document: %w", err)
	}

	return doc, nil
```

PS: due to the deleted line, I cannot add a suggestion for the whole function.

Reply:
```
if !ts.IsZero() {
    doc.Time = ts.Format(TimeFormat)
}
```
I can't factorize this part except by creating a setter that will need one implementation by Document Type. but you're right for the Populate method.

Reply:
Ah in fact, I've to create the setter then I will directly add this part to the specific populate methods

Reply:
Any idea for why do we have to set of Time ? when we extract data with FLBDecoder (`func GetRecord(dec *FLBDecoder) (ret int, ts interface{}, rec map[interface{}]interface{})`), it returns a time and the record data but we set first from the time returned and then a second times with a time from record data, ignoring the first property set.

Reply:
If I am not wrong, we have a lot of timestamp source:
1. From fluentbit: Not sure where does the timestamp come from.
1. From the CRI (Container Runtime Interface): docker adds timestamp to log entries.
1. From the application: Every software logger adds timestamp.

The `ts` variable here is the FluentBit one. It is the default value for `time` field.

Some CRI (containerd) does not add the timestamp to logs entries.

The timestamp from application logger cannot be used here since log entries are not parsed.

Reply:
If I am not wrong, we have a lot of timestamp source:
1. From fluentbit: Not sure where does the timestamp come from.
1. From the CRI (Container Runtime Interface): docker adds timestamp to log entries.
1. From the application: Every software logger adds timestamp.

The `ts` variable here is the FluentBit one. It is the default value for `time` field.

Some CRI (containerd) does not add the timestamp to logs entries.

The timestamp from application logger cannot be used here since log entries are not parsed.

Reply:
Done, I suggest that you take particular attention to `ts`, the fluentbit time record, that I moved into the LogEntry.Populate method where it is useful.
Moreover, by moving the set next to the other, I suspect that the CRI record time also needs a Time Format but I didn´t want to take the initiative without your advice.